File: 0187-repeated-dna-sequences/0187-repeated-dna-sequences.cpp
class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        if(s.size() < 10) return {};
        
        unordered_map<int, int> mp;
        vector<string> res;
        int hash = 0;
        
        unordered_map<char, int> val = {
            {'A', 0}, {'C', 1}, {'G', 2}, {'T', 3}
        };
        
        for(int i = 0; i < s.size(); i++) {
            hash = ((hash << 2) | val[s[i]]) & ((1 << 20) - 1);
            
            if(i >= 9) {
                mp[hash]++;
                if(mp[hash] == 2) {
                    res.push_back(s.substr(i - 9, 10));
                }
            }
        }
        return res;
    }
};

File: 0187-repeated-dna-sequences/0187-repeated-dna-sequences.java
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        List <String> res = new ArrayList<>();
        if(s.length() < 10) return res;
        Map <Integer, Integer> countMap = new HashMap<>();
        Map <Character, Integer> val = new HashMap<>();
        val.put('A', 0);
        val.put('C', 1);
        val.put('G', 2);
        val.put('T', 3);
        int hash = 0;
        int n = s.length();
        for(int i= 0; i< n; i++){
            hash = ((hash << 2)|val.get(s.charAt(i)))&((1<<20)-1); 
            if(i >= 9){
                countMap.put(hash, countMap.getOrDefault(hash, 0) + 1);
                if(countMap.get(hash) == 2){
                    res.add(s.substring(i-9, i+1));
                }
            }
        }
        return res;
    }
}